plugin/pulpcore/plugin/stages/models.py

```diff
@@ -0,0 +1,75 @@
+from gettext import gettext as _
+
+
+class DeclarativeArtifact:
+    """
+    Relates an :class:`~pulpcore.plugin.models.Artifact`, how to download it, and its
+    `relative_path` used later during publishing.
+
+    This is used by the Stages API stages to determine if an
+    :class:`~pulpcore.plugin.models.Artifact` is already present and ensure Pulp can download it in
+    the future. The `artifact` can be either saved or unsaved. If unsaved, the `artifact` attributes
+    may be incomplete because not all digest information can be computed until the
+    :class:`~pulpcore.plugin.models.Artifact` is downloaded.
+
+    Attributes:
+        artifact (:class:`~pulpcore.plugin.models.Artifact`): An
+            :class:`~pulpcore.plugin.models.Artifact` either saved or unsaved. If unsaved, it
+            may have partial digest information attached to it.
+        url (str): the url to fetch the :class:`~pulpcore.plugin.models.Artifact` from.
+        relative_path (str): the relative_path this :class:`~pulpcore.plugin.models.Artifact`
+            should be published at for any Publication.
+        remote (:class:`~pulpcore.plugin.models.Remote`): The remote used to fetch this
+            :class:`~pulpcore.plugin.models.Artifact`.
+
+    Raises:
+        ValueError: If `artifact`, `url`, `relative_path`, or `remote` are not specified.
+    """
+
+    __slots__ = ('artifact', 'url', 'relative_path', 'remote')
+
+    def __init__(self, artifact=None, url=None, relative_path=None, remote=None):
+        if not url:
+            raise ValueError(_("DeclarativeArtifact must have a 'url'"))
+        if not relative_path:
+            raise ValueError(_("DeclarativeArtifact must have a 'relative_path'"))
+        if not remote:
+            raise ValueError(_("DeclarativeArtifact must have a 'remote'"))
+        if not artifact:
+            raise ValueError(_("DeclarativeArtifact must have a 'artifact'"))
```

Comment:
would `assert` be more appropriate?

Reply:
I could see using `assert`, but when I was reading through the standard library docs the [ValueError description](https://docs.python.org/3/library/exceptions.html#ValueError) jumped out at me as describing this error situation perfectly.